src/master-playlist-controller.js

```diff
@@ -831,40 +847,81 @@ export default class MasterPlaylistController extends videojs.EventTarget {
     });
   }
 
-  updateCues_(media) {
+  updateCues_(media, offset = 0) {
     if (!this.useCueTags_ || !media.segments) {
       return;
     }
 
-    while (this.cueTagsTrack_.cues.length) {
-      this.cueTagsTrack_.removeCue(this.cueTagsTrack_.cues[0]);
-    }
-
-    let mediaTime = 0;
+    let mediaTime = offset;
+    let track = this.cueTagsTrack_;
+    let cue;
 
     for (let i = 0; i < media.segments.length; i++) {
       let segment = media.segments[i];
 
-      if ('cueOut' in segment || 'cueOutCont' in segment || 'cueIn' in segment) {
-        let cueJson = {};
+      if (cue !== undefined) {
 
-        if ('cueOut' in segment) {
-          cueJson.cueOut = segment.cueOut;
+        if ('cueIn' in segment) {
+          // Found a CUE-IN so end the cue
+          if (mediaTime !== cue.endTime) {
+            // If the end time does not match the current mediaTime, a CUE-IN may have
+            // been inserted earlier than expected, so the cue should be updated.
+            cue.endTime = mediaTime;
+            cue.adEndTime = mediaTime;
+          }
+          mediaTime += segment.duration;
+          cue = undefined;
+          continue;
         }
-        if ('cueOutCont' in segment) {
-          cueJson.cueOutCont = segment.cueOutCont;
+
+        if (mediaTime < cue.endTime) {
+          // Already processed this mediaTime for this cue
+          mediaTime += segment.duration;
+          continue;
         }
-        if ('cueIn' in segment) {
-          cueJson.cueIn = segment.cueIn;
+
+        // otherwise extend cue until a CUE-IN is found
+        cue.endTime += segment.duration;
+
+      } else {
+        // Since the cues will span for at least the segment duration, adding a fudge
+        // factor of half segment duration will prevent duplicate cues from being
+        // created when timing info is not exact (e.g. cue start time initialized
+        // at 10.006677, but next call mediaTime is 10.003332 )
+        cue = findAdCue(track, mediaTime + (segment.duration / 2));
```

Comment:
The code may be easier to read if, instead of doing the find here and decrementing i to redo the loop, move a conditional check up to do

```
if (!cue) {
  cue = findAdCue(...);
}
```

around the beginning of the loop, then continue with the rest of the code.